src/http/matcher/mod.rs

```diff
@@ -880,6 +696,504 @@ impl<State, Body> crate::service::Matcher<State, Request<Body>> for HttpMatcherK
             HttpMatcherKind::Header(header) => header.matches(ext, ctx, req),
             HttpMatcherKind::Socket(socket) => socket.matches(ext, ctx, req),
             HttpMatcherKind::Any(all) => all.iter().matches_or(ext, ctx, req),
+            HttpMatcherKind::Custom(matcher) => matcher.matches(ext, ctx, req),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use http::{HeaderName, HeaderValue, Version};
+
+    use crate::service::Matcher;
+
+    use super::*;
+
+    struct BooleanMatcher(bool);
+
+    impl Matcher<(), Request<()>> for BooleanMatcher {
+        fn matches(
+            &self,
+            _ext: Option<&mut Extensions>,
+            _ctx: &Context<()>,
+            _req: &Request<()>,
+        ) -> bool {
+            self.0
+        }
+    }
+
+    #[test]
+    fn test_matcher_ands_combination() {
+        let cases = [
```

Comment:
You can add `itertools` as a dev dependency.
This way you can just do:

```rust
use itertools::Itertools;

#[test]
fn test_matcher_ands_combination() {
    for (a, b, c) in [true, false].permutations(3) {
        let expected = a && b && c;
        let a = BooleanMatcher(a);
        let b = BooleanMatcher(b);
        let c = BooleanMatcher(c);

        let matcher = a.and(b).and(c);
        let req = Request::builder().body(()).unwrap();
        assert_eq!(
            matcher.matches(None, &Context::default(), &req),
            expected,
            "({:#?}).matches({:#?})",
            matcher,
            req
        );
    }
}
```

Seems a lot less trouble, and easier to read/maintain?
If you agree you can apply this for all these tests.